### PR TITLE
Remainder operator

### PR DESCRIPTION
**Pull request by nivpgir**

### CODE REVIEW

File: src/compiler/parse.rs
Comment (slightknack) [inline src/compiler/parse.rs:183]: Would it make more sense to call it `Token::Rem` and the function `self.rem(...)` for consistency with the other operators?

File: src/core/math.rs
Comment (slightknack) [inline src/core/math.rs:52]: So here we're using Rust's remainder between floats. There are multiple different ways to do remainders on negative numbers, so I guess we need to figure out whether we prefer the current way remainders are done for negative numbers in Rust or `rem_euclid`, which calculates the least nonnegative remainder?

**Review by slightknack — CHANGES_REQUESTED** [review]
Great work! There are a few small changes I think we should make for consistency with the rest, thanks for taking the time to put in a PR! Note that if you instead decide to merge into `master` rather than `dev`, we can include this feature right now - otherwise, we'll have to wait for the 0.10.0 release.

File: src/compiler/parse.rs
Comment (nivpgir) [reply src/compiler/parse.rs:183]: I kept it as `Percent` in the lexing stage because at that point it's really just a symbol, it gets the meaning of remainder only later, at the parsing stage actually TBH, only at the FFI stage, but ATM we don't use it for anything else, I actually considered changing it to `Percent` in the parsing stage as well, exactly for these reasons, because technically, one could decide to use '%' as an operator for other functions, on other types, but that would be getting ahead of ourselves.

Anyways, the point is I thought it makes sense to differentiate the symbol name when lexing with it's semantic meaning when parsing.

I could change it, your call, let me know.

Comment (slightknack) [reply src/compiler/parse.rs:183]: I don't really think there are any other cases where we use `%` in the language itself. I think the case could be made for something like `-`, because that can be subtraction or unary negation, but `%` itself is pretty much only used for remainder from what I've seen.

File: src/core/math.rs
Comment (nivpgir) [reply src/core/math.rs:52]: I don't have any preferences about that right now. problems with this usually occur when the user expects some behavior (usually coming from a different language) and then gets some different behavior. I usually opt for the mathematical convention in these cases, but I'm not sure what it is for modulus operations... I'll try to think about it today, and we'll see. if you have some preference we could just use that, I don't have strong opinions on this.

File: src/compiler/parse.rs
Comment (nivpgir) [reply src/compiler/parse.rs:183]: I'm mostly using python for work these days, so I see many weird overloads for operators.
a pretty common is using '%' for custom string formatting, if someone won't like the default option (like python, ruby, and probably more).
Of course you could always argue that one _shouldn't_ do that, and that a normal function is better for this use case, and that '%' should be used only as the `remainder` operator, I guess since operators are actually implemented with macros + FFI, it sounds pretty common for me that someone would want to override it for some DSL or something, but maybe I'm already thinking too far ahead...?
Personally I like keeping my options open, but I get that it could also lead to a mess, so IDK.

File: src/core/math.rs
Comment (slightknack) [reply src/core/math.rs:52]: I'm pretty sure the mathematical convention for modulus operators is the euclidean remainder, which is always positive. I was a bit confused with rust's default remainder behavior when I first came across it. Let me know what you think.

File: src/compiler/parse.rs
Comment (slightknack) [reply src/compiler/parse.rs:183]: Here are my thoughts: I think that % string formatting isn't ideal, and much prefer the f-string formatting present in languages like Python, and Rust. I think that we'll eventually support overloading for all operators, but this won't be done though implementing something for `%` directly, rather you'd have to implement a `rem` trait, something like `__rem__` in Python (actual traits though, not just dunder methods on classes.)

**Comment by slightknack** [comment]
By the way, once this pull request is merged I'll bump the version to 0.9.1 and publish the release on crates.

File: src/compiler/parse.rs
Comment (nivpgir) [reply src/compiler/parse.rs:183]: Ok, I have a great ability to over-discuss the _too small_ details, and I feel like we're getting to that point so I'll try to avoid it.
I think we're not at the point where we can accurately know how to work on the things I brought up here, traits sound better and safer to me, but we have a few bridges to cross before actually getting there, right?
So bottom line:
I like the separation in names between the token (percent) and it's semantic meaning (remainder), regardless of whether it will be overloaded or not, I like it like this, I don't find it confusing, I actually think it's clearer, but if you don't agree, I'll change it, it's not a big deal.

File: src/core/math.rs
Comment (nivpgir) [reply src/core/math.rs:52]: Sure, I'll change it to use `rem_euclid`

File: src/compiler/parse.rs
Comment (slightknack) [reply src/compiler/parse.rs:183]: Thank you. If it ever becomes clear that `%` will be used for things other than remainder going forward, we can change the name of the token variant, no biggie 😄

**Comment by nivpgir** [comment]
changes were made, we're gonna push this to master instead

**Pull request closed without merging**
